### PR TITLE
Feature/gitlab integration

### LINKED ISSUE #11104: Gitlab pipelines build plugin
### Feature description

I'd like to implement and contribute this plugin. At this moment I have a working prototype in my personal repo.
https://github.com/e-stadnik/gitextensions/tree/GitlabIntegration
Will be happy for any feedback and suggestions.

### Environment

- Git Extensions 3.5.4.12724
- Build 65f01f39982ccb121c4df7d4b00b506939553958
- Git 2.33.0.windows.2
- Microsoft Windows NT 10.0.19045.0
- .NET Framework 4.8.9167.0
- DPI 144dpi (150% scaling)

### PR DESCRIPTION
**Pull request by e-stadnik**
Fixes #11104

## Proposed changes

- Added build server integration for Gitlab pipelines 

### After

image

## Test methodology

- Unit tests for adapter logic are included
- Manual testing on private commercial solution on self-hosted Gitlab instance
- Manual testing on open-source repo: https://gitlab.com/wireshark/wireshark

## Test environment(s)

- GIT 2.33.0.windows.2
- Windows 10 22H2 19045.3208

## Merge strategy
- To be decided later.

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by gerhardol** [comment]
Can this be tried with public gitlab.com?
What to configure?
Can you please add this to [source/settings.rst](https://github.com/gitextensions/GitExtensionsDoc/blob/main/source/settings.rst)?

File: GitExtensions.sln [hunk 1]
```diff
@@ -15,28 +15,29 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Statistics", "Statistics",
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "GitExtensions", "GitExtensions\GitExtensions.csproj", "{F15A69AF-7EBD-4F69-A026-5071FA3EC61B}"
 	ProjectSection(ProjectDependencies) = postProject
+		{0B131F9D-BB10-4C3E-9656-5AB858158DE5} = {0B131F9D-BB10-4C3E-9656-5AB858158DE5}
```
Comment (gerhardol) [inline GitExtensions.sln:18]: Please revert the non relevant changes to this file

File: Plugins/BuildServerIntegration/GitlabIntegration/ApiClient/GitlabApiClient.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            _projectId = projectId;
+        }
+
+        public async Task<PagedResponse<GitlabPipeline>> GetPipelinesAsync(DateTime? sinceDate, bool running, int pageNumber)
+        {
+            UriBuilder pipelinesUriBuilder = new($"{InstanceUrl}/api/v4/projects/{_projectId}/pipelines");
+            NameValueCollection query = HttpUtility.ParseQueryString(pipelinesUriBuilder.Query);
+
+            if (sinceDate != null)
```
Comment (gerhardol) [inline Plugins/BuildServerIntegration/GitlabIntegration/ApiClient/GitlabApiClient.cs:29]:
```suggestion
            if (sinceDate is not null)
```

**Review by gerhardol — COMMENTED** [review]
textboxes are wider than the the minimal width
very brief review but did not react to anything

**Comment by RussKie** [comment]
Thank you for the contribution but I am really not excited about adding
another plugin in to the main repo.

My lack of excitement stems from several things:
1. Ownership - the Core team will essentially be responsible for
maintaining the plugin, triaging issues and fixing bugs after a short
while.

2. Release cadence - we're releasing the app few times a year,  so any
issues with the plugin won't get addressed in timely manner. E.g., GitLab
changed its API, or bugs discovered in the plugin or few features
requested.


It'd be desirable if the plugin was released as a standalone plugin which
users can install via the Plugin Manager.

**Comment by e-stadnik** [comment]
Thank you for feedback and review tips.

As Gitlab mentioned in [docs](https://docs.gitlab.com/ee/api/rest/#compatibility-guideline) all significant API changes come major GitLab releases only. It should be very minor chance to receive unexpected failtures.

The logic here is quite simple and similar to other integrations included in main repo.

Can we discuss durability requirements which will make possible to merge it in product?
Because Gitlab.com allows token free access for public repos, I'm about to add integration tests.

**Comment by gerhardol** [comment]
> The logic here is quite simple and similar to other integrations included in main repo.
> 
> Can we discuss durability requirements which will make possible to merge it in product? Because Gitlab.com allows token free access for public repos, I'm about to add integration tests.

The difference is if any maintainer use the feature or not. (I have no info if someone uses TeamCity or if the integration exists). It is quite frustrating to get issue reports for features you know nothing about and cannot test.

Build server integration are quite special plugins, deeply integrated to GE so there are arguments to keep them in as well, then general cleanups are done as well.
@RussKie takes the decision.

**Review by gerhardol — COMMENTED** [review]
Still a visual:
image

Can you please describe how to use? 
Can it be used with gitlab.com?

**Comment by e-stadnik** [comment]
> Can you please describe how to use?
> Can it be used with gitlab.com?

Sure, I tested on [wireshark repo](https://gitlab.com/wireshark/wireshark)

For settings use Instance URL https://gitlab.com/
And project id 7898047
Token field can be left empty because it's a public project.

For private projects token can be obtained according to this [guide](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#create-a-personal-access-token)

**Comment by pmiossec** [comment]
Some potential improvements:

> > Can it be used with gitlab.com?
> 
> Sure, I tested on [wireshark repo](https://gitlab.com/wireshark/wireshark)

So maybe it could be a good idea to set "https://gitlab.com" by default

> For settings use Instance URL https://gitlab.com/ And project id 7898047 Token field can be left empty because it's a public project.

If projectId is mandatory but not the token, for me it makes more sense to put the projectId field before the token one.


> For private projects token can be obtained according to this [guide](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#create-a-personal-access-token)

I would at least put a link to this documentation. But even better looking at the documentation, you can build a link open the url (using the server url) to let the user easily create a token. From the doc `https://gitlab.example.com/-/profile/personal_access_tokens?name=GitExtensions+Access+token&scopes=api,read_user,read_registry`...
It is already done for the AzureDevOps Ci (`TokenManagementLink` control in `SettingsUserControl.cs`)

**Comment by e-stadnik** [comment]
Philippe, thanks for your feedback!

I'm going to fill URL and project ID right from remotes.
Along with a link to obtain a token if necessary.

**Comment by pmiossec** [comment]
> I'm going to fill URL and project ID right from remotes.

Great!
But is it possible?
I don't know GitLab works but previously you said that you need the projectId `7898047` but you don't have it from the remote url ( `https://gitlab.com/wireshark/wireshark.git`).... There is a open api to get it?

**Comment by e-stadnik** [comment]
> But is it possible?

I'll try to lookup this way https://docs.gitlab.com/ee/api/projects.html#search-for-projects-by-name
When url_to_repo matches with remote it will be exact project id

**Comment by RussKie** [comment]
I'm away on holidays without an access to a computer (digital detox 😆). So
it may be some time before I can check it out.

**Comment by pmiossec** [comment]
> digital detox 😆

Mine starting in 2 days 😆

**Comment by e-stadnik** [comment]
> It's a little strange but with the request https://gitlab.com/api/v4/projects?search=wireshark&sort=asc I don't get the wireshark project in the list returned. We need a token to have the full list?

You see paged response. The target one can be somewhere deep.
But I've found other issue, that project can be renamed without changing remote path.

**Comment by pmiossec** [comment]
> But I've found other issue, that project can be renamed without changing remote path.

Yes, so don't add too much complexity for something that is only made once and takes 2 minutes to do.
Even me that don't know gitlab found it in 10s...

**Comment by e-stadnik** [comment]
> > But I've found other issue, that project can be renamed without changing remote path.
> 
> Yes, so don't add too much complexity for something that is only made once and takes 2 minutes to do. Even me that don't know gitlab found it in 10s...

The solutions was [easier](https://docs.gitlab.com/ee/api/rest/index.html#namespaced-path-encoding)
https://gitlab.com/api/v4/projects/wireshark%2Fwireshark returns whole project info.

**Comment by gerhardol** [comment]
The ProjectId seems to be set for only one of the remotes. In my case I have "origin" with my changes and "upstream". The status is only shown for "origin".
AppVeyor and Jenkins at least allows several urls.

In addition, ProjectId is normally empty when opening the settings. I would like to say that the value is not read too, but it may be that I only got the build for origin but were looking for upstream builds.

**Comment by e-stadnik** [comment]
I've never worked in multiple remotes with configured CI/CD on different servers.
But think Gitlab is mostly designed to operate with stages in a single project.
The stages are equivalent of projects in TeamCity terminology.

Maybe I miss some important case, but anyway to support pipeline statuses from multiple origins we should allow to set several instance URLs.

About ProjectId loading, it was bug, fixed :)

**Comment by gerhardol** [comment]
Saving ProjectId seem to work for me now.

My question was about multiple owners on the same server, origin and upstream.
It is at least possible to switch between them now, so you can see builds from each even if it is hard to switch to upstream.
Not necessarily something to handle now.

I actually use multiple servers too. Not just different URL but different implementations for different suppliers: Gitlab to BitBucket to GitHub and back again.
But that is a separate setup, I use unique clones to handle this. Can unfortunately not use WorkTrees, GE does not store unique data per repo (not a big problem, build server is only important for me in BB). No private clones in BB, but multiple 'users' in GL.

@RussKie How to proceed here? Include this now and separate later?
If the plugin fails, it will have to wait for next normal update...

**Comment by spdr870** [comment]
@e-stadnik 
I use the gitlab community edition, self-hosted. I tried this plugin. The status is visible. The biggest issue I have is the performance. The repository is over 200.000 commits, and the UI freezes until the status is fetched for, I assume, all commits? To make it workable, I changed the code only to load the first page. Not sure how other plugins work, but a limit would be needed for large repositories. And maybe check for `cancellationToken.IsCancellationRequested` when loading pages?

Also; the build report tab page couldn't load the content since the inline browser is not supported by gitlab. Not a problem for me, the link to open it in an external browser works.

**Comment by e-stadnik** [comment]
@spdr870, Thanks a lot for you feedback!

Can you see page numbers on pipelines web page for you project?
From some threshold plugin loading all pages one by one, because total pages disappears from API as well. (like in wireshark repo). I didn't face with performance issues here.

If value for you project is below, plugin trying to request pages in parallel. In that case I'll try to change scheduler logic and it should helps with performance.

As constant solution I think to cache loaded entities on disk because API will return updated entities.

**Comment by gerhardol** [comment]
> The biggest issue I have is the performance. The repository is over 200.000 commits, and the UI freezes until the status is fetched for, I assume, all commits? To make it workable, I changed the code only to load the first page. Not sure how other plugins work, but a limit would be needed for large repositories. And maybe check for `cancellationToken.IsCancellationRequested` when loading pages?

Build status should be fetched and updated in the background (it is for at least Jenkins and AppVeyor). 

> Also; the build report tab page couldn't load the content since the inline browser is not supported by gitlab. Not a problem for me, the link to open it in an external browser works.

The tab should be disabled by default for all build servers.
The WebBrowser control could be replaced with WebView2 (https://learn.microsoft.com/en-us/microsoft-edge/webview2/) (only Win10 and later) or some Chromium control (adds to size), but I propose to just change the defaults.

**Comment by e-stadnik** [comment]
Hi @spdr870 I'd like to ask you a bit more details regarding performace.
For sure the data are loading in a background, but the callbacks are executing for each commit one by one. I suppose it can cause a freezes if we have small gap between page requests.
Is the Gitlab server in your local network?

It's just a hypothesis can't reproduce it on my environment. I checked performance on [wireshark ](https://gitlab.com/wireshark/wireshark) repo with 10000+ pipelines history and UI responsibility was good enough, but it takes about 300ms to fetch single page data from my location.

We can validate it if you put Thread.Sleep() for 100-300ms in 
[Plugins/BuildServerIntegration/GitlabIntegration/GitlabAdapter.cs](https://github.com/gitextensions/gitextensions/pull/11105/files#diff-b4eef118e95f69c867f255f5eb785a8d9d9c9b56b4011da7b51c59b54889c25a) after line 109

**Comment by e-stadnik** [comment]
I added BuildServer.Gitlab.PagesLimit settings but not sure about showing it in UI
It will appear in GitExtensions.settings file with default value (0 - no limitation) after save action in Gitlab settings page.
@spdr870 can you please check performance now with 10 or 50 pages?

**Comment by e-stadnik** [comment]
> Can this be tried with public gitlab.com? What to configure? Can you please add this to [source/settings.rst](https://github.com/gitextensions/GitExtensionsDoc/blob/main/source/settings.rst)?

UI fields documentation added here: https://github.com/gitextensions/GitExtensionsDoc/pull/148

**Comment by gerhardol** [comment]
@RussKie I suggest that this plugin is included. Separating buildserver plugins may be done later.

**Comment by RussKie** [comment]
Here's a test artifact: https://ci.appveyor.com/project/gitextensions/gitextensions/builds/48079386/artifacts
Could you please take it for a spin, install it and see if everything is working as you expect? If everything is good, we'll merge this.
Thank you.

**Comment by e-stadnik** [comment]
I'd been testing build [4.2.0.17315](https://ci.appveyor.com/project/gitextensions/gitextensions/builds/47929505) for a last 2 weeks on a private repo.

On recent build checked tabIndex and pages loading limitation. Everything looks fine.

**Comment by gerhardol** [comment]
I have been using this pr for some time too, but my interaction with gitlab is quite limited, so it does not push any limit.
No proper review, but it looks good.

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
Thank you

**Pull request merged**